### ISSUE TITLE
Feature Request: Disallow Map<K, V> access via Square Bracket notation

### ISSUE THREAD

**Issue by josephliccini** [palantir/tslint#3448]
This is not a bug report, but a feature request.

We have a lot of C# .NET developers, and they are now doing some more work on the client codebase, using TypeScript.  Since in C# one can access `Dictionary<K, V>` via square bracket notation like:

```cs
var myDict = new Dictionary<string, string>();
var myValue = myDict["foo"];
myDict["bar"] = "something";
```

This syntax is also legal in JavaScript / TypeScript, but is incorrect regarding the `Map<K, V>` interface / implementation.

#### TypeScript code being linted

```ts
const myMap = new Map<string, string>();
myMap['foo'] = 'bar'; // Can we throw a Lint error?
var myValue = myMap['something']; // Also lint error
```

Thoughts?

Thanks!
Joe

**Comment by ajafff** [comment]
Have you tried the `no-string-literal` rule? It forbids the square bracket notation where not strictly necessary. After fixing the lint error, the typescript compiler should show an appropriate error mesage.

```ts
myMap["foo"]; // lint error
// will be fixed to
myMap.foo; // compiler error
```

**Comment by josephliccini** [comment]
Got it.  Yeah I suppose that would work.  Let me give this a try and see if there's any issues.  Sorry for the noise, I thought I checked through all the rules on the website, I must have missed this one.

**Comment by josephliccini** [comment]
@ajafff This helps in many cases, but not all.

It didn't work this case:
```ts
const bar = new Map();
bar[foo] = 'hi'; // No Lint or Compiler error
bar['something'] = 'bye'; // Lint Error
```

**Comment by ajafff** [comment]
@josephliccini have you tried the `--noImplicitAny` compiler option? Is should show an error on all those cases.
